tree/bst.go

```diff
@@ -160,3 +160,117 @@ func (n *node) getBSTNodeWithValue(val int) *node {
 
 	return n.right.getBSTNodeWithValue(val)
 }
+
+// changes By Prerna 08/01/2022
+
+// 1. Is validate BST
+func (t *BinarySearchTree) ValidBST() {
+	if t.isBST() {
+		fmt.Println("Valid BST")
+
+	} else {
+		fmt.Println("Invalid BST")
+	}
+}
+
+func (t *BinarySearchTree) isBST() bool {
+	if t == nil {
+		return true
+	}
+	return t.root.chkNode(100, 100)
+}
+func (n *node) chkNode(max int, min int) bool {
+	if n == nil {
+		return true
+	}
+	if n.val > max && n.val < min {
+		return false
+	}
+	if n.left.chkNode(n.val, min) && n.right.chkNode(max, n.val) {
+		return true
+	}
+	return false
+
+}
+
+// 2. determine if the given BST is balance
+
+func (t *BinarySearchTree) IsBal() {
+	if t == nil {
+		fmt.Println("Balanced")
+		return
+	}
+	h := 0
+	h1 := t.root.left.GetHght(h)
+	h2 := t.root.right.GetHght(h)
+	if h1-h2 > 1 || h2-h1 > 1 {
+		fmt.Println("Un-balanced")
+		return
+	}
+	fmt.Println("Balanced")
+}
+
+func (n *node) GetHght(h int) int {
+	if n == nil {
+		return h
+	}
+	n.left.GetHght(h + 1)
+	n.right.GetHght(h + 1)
+	if n.left.GetHght(h+1) > n.right.GetHght(h+1) {
+		return n.left.GetHght(h + 1)
+	} else {
+		return n.right.GetHght(h + 1)
+	}
+
+}
+
+// 3. given a sorted array, construct a bst with minimal height
+// note: following mthods has error
+
+func ConstructBST(a []int) *BinarySearchTree {
```

Comment:
Let's get this working 💯

Reply:
Remove this function from this branch. Put this CL in separate branch and when it works as expected, create a separate merge/pull request for that. All the rest changes are good and we can merge it.